runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -974,9 +974,12 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
       // If the array is known to have a component type that is not a primitive value type or
       // the value being stored is known not to be a value type, transform the helper
       // call to a regular aaload or aastore
+      // If the value being stored is NULL and the destination array component is null restricted in runtime,
+      // a NPE is expected to throw. In this case, the helper call cannot be transformed because a null check
+      // will be required based on the destination array component type in runtime.
       bool canTransformIdentityArrayElementLoadStore = false;
       if ((arrayConstraint != NULL && isCompTypePrimVT == TR_no)
-          || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
+          || (isStoreFlattenableArrayElement && (isStoreValueVT == TR_no) && storeValueConstraint && !storeValueConstraint->isNullObject()))
```

Comment:
[Code that appears later in this method](https://github.com/eclipse-openj9/openj9/blob/69f5bd9facb053852ddf1cebcb0576fc9260b0a6/runtime/compiler/optimizer/J9ValuePropagation.cpp#L1188-L1193) attempts to ensure that a run-time check will be inserted if it's possible that a null reference will be assigned to an array whose component type is not known to be a non-null restricted type.  Can you explain why that code doesn't handle this situation?  I'm wondering whether it should be possible to handle it there.

Reply:
The [referred code at the above](https://github.com/eclipse-openj9/openj9/blob/69f5bd9facb053852ddf1cebcb0576fc9260b0a6/runtime/compiler/optimizer/J9ValuePropagation.cpp#L1188-L1193) is only enabled if `canTransformUnflattenedArrayElementLoadStore` is true. It would cause a null check being inserted if
(1) Array flattening is disabled
OR
(2) Array flattening is enabled and the destination array is null restricted and not flattened in compilation time.

`canTransformIdentityArrayElementLoadStore` covers the case where array flattening is enabled
(1) The destination array is identity array
OR
(2) The destination array type may or may not be null restricted, but the store value is not null restricted value type (can be null) 

My initial reasoning is that if we set `canTransformIdentityArrayElementLoadStore` as true [as it is today](https://github.com/eclipse-openj9/openj9/blob/69f5bd9facb053852ddf1cebcb0576fc9260b0a6/runtime/compiler/optimizer/J9ValuePropagation.cpp#L982C11-L982C11) to allow the helper call transformation and rely on the [later code](https://github.com/eclipse-openj9/openj9/blob/69f5bd9facb053852ddf1cebcb0576fc9260b0a6/runtime/compiler/optimizer/J9ValuePropagation.cpp#L1188-L1193) to decide whether or not to insert a null check, an unnecessary null check could be added for the case where the destination array is identity array in runtime.

Now on a second thought, I think even with the existing code, inevitably we already introduce an unnecessary null check  
if the destination array is identity array in runtime. For example the case where `canTransformUnflattenedArrayElementLoadStore` is true + the destination array component type may or may not be null restricted in compilation time + `storeValueConstraint` is NULL. In runtime, the destination array component type could end up being an identity type. 

I think an alternative fix is as suggested at the above to allow helper call transformation and let the [later code](https://github.com/eclipse-openj9/openj9/blob/69f5bd9facb053852ddf1cebcb0576fc9260b0a6/runtime/compiler/optimizer/J9ValuePropagation.cpp#L1188-L1193) to decide if a null value check is required. The tradeoff in these two fixes is between not transforming the call at all, or transforming the call with a potential unnecessary null check. Performance wise, I think allowing the transformation with potential extra null check might still be better than not transforming the call at all. @hzongaro What do you think?

Reply:
Thanks, Annabelle @a7ehuo.  Yes, that's what I was thinking - if the array turns out at run-time to have a null-restricted component type, a `NullPointerException` will be thrown; if it's not, the inlined code for the array element assignment can be run safely.  The performance should be better than would result from preventing the transformation.

Reply:
Addressed in e050899. Ready for another view. Thanks!